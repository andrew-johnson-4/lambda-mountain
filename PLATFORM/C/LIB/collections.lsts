
interface Collection<t> {
   let .length(x: self): U64;
   let $"[]"(x: self, idx: U64): t;
};

let .last(self: t+Collection<x>): x = (
   let ln = self.length();
   if ln == 0 {
      fail("cannot get .last() from empty element");
   };
   self[(ln-1)]
);

interface Iterator<t> {
   let .next(x: self[]): Maybe<t>;
};

interface Iterable<t> {
   ## When modifying th eunderlying collection while iterating,
   ## the iterator should continue on the old version.
   let .iter(x: self): Iterator<t>;

   ## Underlying collection not allowed to be modified while iterating.
   let .unsafe-iter(x: self): Iterator<t>;
};

type Iterator<t> implements Iterable<t>;
let .iter(x: t+Iterator<elt>): t = ( x );
let .unsafe-iter(x: t+Iterator<elt>): t = ( x );

interface IteratorRemaining {
   let .remaining(x: self + Iterator<t>): U64;
};

let remaining-or-zero(from: Iterator<x>): U64 = (
   0
);

let remaining-or-zero(from: Iterator<x>+IteratorRemaining): U64 = (
   from.remaining()
);

let length-or-zero(from: Iterable<x>): U64 = (
   0_u64
);

let length-or-zero(from: Collection<x>): U64 = (
   from.length()
);

let length-or-zero(from: Collection<x>+Iterable<x>): U64 = (
   from.length()
);

type CollectionIteratorBase<c,t> = CollectionIteratorBase { coll: c, idx: U64, _u: Unused<t> };
type CollectionIteratorBase<c,t> implements Iterator<t>;
type CollectionIteratorBase<c,t> implements IteratorRemaining;

let .remaining(self: CollectionIteratorBase<c,t>): U64 = (
   self.coll.length() - self.idx
);

let .next(p: CollectionIteratorBase<c,t>[]): Maybe<t> = (
   let v = p[0];
   if v.idx < v.coll.length() {
      let val = v.coll[v.idx];
      v.idx = v.idx + 1;
      p[0] = v;
      Some { val } :: Maybe<t>
   } else {
      None :: Maybe<t>
   }
);

let .into(self: c+Collection<t>, out: Type<CollectionIteratorBase<c,t>>): CollectionIteratorBase<c,t> = (
   CollectionIteratorBase { self, 0_u64, unused(type(t)) }
);
