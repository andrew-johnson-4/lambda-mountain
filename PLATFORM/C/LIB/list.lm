
cons := λ(: hd x)(: tl List<x>). (: (
   (set tl (LCons( hd (close tl) )))
   tl
) List<x>);

list::cons := λ(: hd x)(: tl List<x>). (: (
   (set tl (LCons( hd (close tl) )))
   tl
) List<x>);

head := λ(: l List<x>). (: (
   (match l (
      ()
      ( LEOF (fail 'List::head\sis\sfallible\n_s) )
      ( _ () )
   ))
   (.2(as l List<x>+Tag::LCons))
) x);

.has-head := λ(: l List<x>). (: (
   (match l (
      ()
      ( LEOF 0_u64 )
      ( _ 1_u64 )
   ))
) U64);

tail := λ(: l List<x>). (: (
   (match l (
      ()
      ( LEOF (fail 'List::head\sis\sfallible\n_s) )
      ( (LCons( hd tl ))  (set l tl) )
   ))
   l
) List<x>);

+ := λ(: hd List<x>)(: tl List<x>). (: (
   (set hd (.reverse hd))
   (while (non-zero hd) (match hd (
      ()
      ( (LCons( hdhd tlhd )) (
         (set tl (cons( hdhd tl )))
         (set hd tlhd)
      ))
   )))
   tl
) List<x>);

.length := λ(: list List<x>). (: (
   (let l 0_u64)
   (while (non-zero list) (
      (match list (
         ()
         ( (LCons( _ tl )) (set list tl) )
      ))
      (set l (+( l 1_u64 )))
   ))
   l
) U64);

.reverse := λ(: l List<x>). (: (
   (let r (: LEOF List<x>))
   (while (non-zero l) (match l (
      ()
      ( (LCons( v rst )) (
         (set r (LCons( v (close r) )))
         (set l rst)
      ))
   )))
   r
) List<x>);

print := λ(: l List<x>). (: (
   (print '[_s)
   (let some False_u8)
   (while (non-zero l) (match l (
      ()
      ( (LCons( v rst )) (
         (if (==( some True_u8 )) (print ',_s) ())
         (print v)
         (set some True_u8)
         (set l rst)
      ))
   )))
   (print ']_s)
) Nil);

== := λ(: lt List<x>)(: rt List<x>). (: (
   (let r 0_u64)
   (match (Tuple( lt rt )) (
      ()
      ( (Tuple( LEOF LEOF )) (set r 1_u64) )
      ( (Tuple( (LCons( ltr lt1 )) (LCons( rtr rt1 )) )) (
         (if (==( lt1 rt1 )) (
            (if (==( ltr rtr )) (
               (set r 1_u64)
            ) ())
         ) ())
      ))
      ( _ () )
   ))
   r
) U64);

.contains := λ(: l List<x>)(: i x). (: (
   (let r 0_u64)
   (while (non-zero l) (
      (if (==( (head l) i )) (set r 1_u64) ())
      (set l (tail l))
   ))
   r
) U64);

.join :=  λ(: l List<String>)(: sep String). (: (
   (let r (SNil()))
   (while (non-zero l) (
      (set r (+( r (SAtom(head l)) )))
      (set l (tail l))
      (if (non-zero l) (set r (+( r (SAtom sep) ))) ())
   ))
   (clone-rope r)
) String);

