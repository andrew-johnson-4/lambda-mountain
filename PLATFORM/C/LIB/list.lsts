
## linked list
##
## consider using [type(Vector<x>)] instead
type List<x> zero LEOF
             = LEOF
             | LCons{ head:x , tail:List<x>[] };

type List<x> implements Iterable<x>;
type List<x> implements Collection<x>;

let cmp(ls: List<x>, rs: List<x>): Ord = (
   let diff = Equal();
   while is(diff, Equal()) && non-zero(ls) && non-zero(rs) {
      diff = cmp(head(ls), head(rs));
      ls = tail(ls);
      rs = tail(rs);
   };
   if not(is(diff,Equal())) then diff else
   if not(non-zero(ls)) && non-zero(rs) then LessThan() else
   if non-zero(ls) && not(non-zero(rs)) then GreaterThan() else
   Equal()
);

let to-smart-string(ls: List<x>): String = (
   let s = "[";
   let si = 0_u64;
   for l in ls {
      if si > 0_u64 {
         s = s + ",";
      };
      s = s + to-smart-string(l);
      si = si + 1_u64;
   };
   s = s + "]";
   s
);

let .unique(ls: List<x>): List<x> = (
   let rs = [] : List<x>;
   for l in ls {
      if not(rs.contains(l)) {
         rs = cons(l, rs);
      }
   };
   rs
);

##[DEPRECATED .iter().next()]
let .next(ls: Array<List<x>,?>): Maybe<x> = (
   match open(ls) {
      [i.. rst] => (
         ls[0_u64] = rst;
         Some(i)
      );
      [] => (None : Maybe<x>)();
   };
);

let .nth(tt: List<x>, idx: U64): Maybe<x> = (
   let err = false;
   while idx > 0 && not(err) {
      if tt.has-head() {
         tt = tail(tt);
      } else {
         err = true;
      }
   };

   if err {
      None : Maybe<x>
   } else {
      Some(head(tt))
   }
);


let $"[]"(tt: List<x>, idx: U64): x = (
   while idx > 0 {
      if not(tt.has-head()) {
         fail("index \{idx} is out of bounds for list of length \{tt.length()}");
      };
      tt = tail(tt);
   };
   head(tt)
);

type ListIter<x> implements Iterator<x> = { list: List<x> };

let .next(ls: ListIter<x>[]): Maybe<x> = (
   match open(open(ls).list) {
      [i.. rst] => (
         ls[0_u64] = ListIter(rst);
         Some(i)
      );
      [] => (None : Maybe<x>);
   };
);

let .iter(li: List<x>): ListIter<x> = (
   li.unsafe-iter();
);

let .unsafe-iter(li: List<x>): ListIter<x> = (
   ListIter(li)
);

let .lookup(ls: List<Tuple<k,v>>, key: k, default: v): v = (
   let found = false;
   for Tuple{lkey=first, rval=second} in ls {
      if not(found) && key==lkey { default=rval; found=true; }
   };
   default;
);
let .lookup(ls: List<Tuple<k,v1,v2>>, key: k, default: Tuple<v1,v2>): Tuple<v1,v2> = (
   let found = false;
   for Tuple{lkey=first, val1=second, val2=third} in ls {
      if not(found) && key==lkey { default=Tuple(val1,val2); found=true; }
   };
   default;
);

