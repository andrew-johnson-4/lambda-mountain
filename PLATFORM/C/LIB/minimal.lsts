
import PLATFORM/C/LIB/primitive-types.lsts;
import PLATFORM/C/LIB/blob.lsts;
import PLATFORM/C/LIB/c-shim.lm;
import PLATFORM/C/LIB/common-macros.lm;
import PLATFORM/C/LIB/common-macros.lsts;
import PLATFORM/C/LIB/platform-macros.lm;
import PLATFORM/C/LIB/blob.lm;
import PLATFORM/C/LIB/primitives.lm;
import PLATFORM/C/LIB/primitives.lsts;
import PLATFORM/C/LIB/field-accessors.lsts;
import PLATFORM/C/LIB/sized.lm;
import PLATFORM/C/LIB/bool.lsts;
import PLATFORM/C/LIB/i64.lm;
import PLATFORM/C/LIB/u64.lm;
import PLATFORM/C/LIB/u8.lsts;
import PLATFORM/C/LIB/u16.lsts;
import PLATFORM/C/LIB/u32.lsts;
import PLATFORM/C/LIB/u64.lsts;
import PLATFORM/C/LIB/i32.lsts;
import PLATFORM/C/LIB/i64.lsts;
import PLATFORM/C/LIB/usize.lsts;
import PLATFORM/C/LIB/string.lm;
import PLATFORM/C/LIB/string.lsts;
import PLATFORM/C/LIB/smart-string.lm;
import PLATFORM/C/LIB/smart-string.lsts;
import PLATFORM/C/LIB/s.lsts;
import PLATFORM/C/LIB/tuple.lsts;
import PLATFORM/C/LIB/list.lm;
import PLATFORM/C/LIB/list.lsts;
import PLATFORM/C/LIB/maybe.lsts;
import PLATFORM/C/LIB/vector.lsts;
import PLATFORM/C/LIB/hashtable.lm;
import PLATFORM/C/LIB/hashtable.lsts;
import PLATFORM/C/LIB/array.lm;
import PLATFORM/C/LIB/array.lsts;
import PLATFORM/C/LIB/io.lm;
import PLATFORM/C/LIB/io.lsts;
import PLATFORM/C/LIB/regex.lm;
import PLATFORM/C/LIB/regex.lsts;
import PLATFORM/C/LIB/cmp.lsts;
import PLATFORM/C/LIB/print.lsts;
