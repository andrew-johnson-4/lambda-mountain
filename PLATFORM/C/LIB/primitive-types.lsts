
type U8 implies U16, I16, U32, I32, U64, I64 implements DefaultPrintable suffix _u8;
type I8 implies I16, I32, I64 implements DefaultPrintable suffix _i8;
type U16 implies U32, I32, U64, I64 implements DefaultPrintable suffix _u16;
type I16 implies I32, I64 implements DefaultPrintable suffix _i16;
type U32 implies U64, I64 implements DefaultPrintable suffix _u32;
type I32 implies I64 implements DefaultPrintable suffix _i32;
type U64 implements DefaultPrintable suffix _u64;
type I64 implements DefaultPrintable suffix _i64;
type F64 suffix _f64;

type String implements Collection<U8> suffix _s;
type SmartString implements Collection<U8> suffix _ss
                = { data:U8[] , start:U8[] , end:U8[] , data-end:U8[] };
