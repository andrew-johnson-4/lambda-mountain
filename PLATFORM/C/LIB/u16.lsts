
declare-binop( $"!=", raw-type(U16), raw-type(U16), raw-type(U64), ( l"("; x; l"!="; y; l")"; ) );
declare-binop( $"==", raw-type(U16), raw-type(U16), raw-type(U64), ( l"("; x; l"=="; y; l")"; ) );
declare-binop( $"<",  raw-type(U16), raw-type(U16), raw-type(U64), ( l"("; x; l"<"; y; l")"; ) );
declare-binop( $"<=", raw-type(U16), raw-type(U16), raw-type(U64), ( l"("; x; l"<="; y; l")"; ) );
declare-binop( $">",  raw-type(U16), raw-type(U16), raw-type(U64), ( l"("; x; l">"; y; l")"; ) );
declare-binop( $">=", raw-type(U16), raw-type(U16), raw-type(U64), ( l"("; x; l">="; y; l")"; ) );

declare-binop( $"+", raw-type(U16), raw-type(U16), raw-type(U16), ( l"("; x; l"+"; y; l")"; ) );
declare-binop( $"-", raw-type(U16), raw-type(U16), raw-type(U16), ( l"("; x; l"-"; y; l")"; ) );
declare-binop( $"^", raw-type(U16), raw-type(U16), raw-type(U16), ( l"("; x; l"^"; y; l")"; ) );
declare-binop( $"/", raw-type(U16), raw-type(U16), raw-type(U16), ( l"("; x; l"/"; y; l")"; ) );
declare-binop( $"%", raw-type(U16), raw-type(U16), raw-type(U16), ( l"("; x; l"%"; y; l")"; ) );
declare-binop( $"*", raw-type(U16), raw-type(U16), raw-type(U16), ( l"("; x; l"*"; y; l")"; ) );

declare-unop( not, raw-type(U16), raw-type(U64), ( l"(!"; x; l")"; ) );

declare-unop( into-branch-conditional, raw-type(U16), raw-type(BranchConditional), x );

let print(io: IO::File, x: U16): Nil = (fprintf(io, c"%u", x); ());
