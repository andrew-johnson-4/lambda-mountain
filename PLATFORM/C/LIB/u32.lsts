
declare-binop( $"!=", raw-type(U32), raw-type(U32), raw-type(U64), ( l"("; x; l"!="; y; l")"; ) );
declare-binop( $"==", raw-type(U32), raw-type(U32), raw-type(U64), ( l"("; x; l"=="; y; l")"; ) );
declare-binop( $"<",  raw-type(U32), raw-type(U32), raw-type(U64), ( l"("; x; l"<"; y; l")"; ) );
declare-binop( $"<=", raw-type(U32), raw-type(U32), raw-type(U64), ( l"("; x; l"<="; y; l")"; ) );
declare-binop( $">",  raw-type(U32), raw-type(U32), raw-type(U64), ( l"("; x; l">"; y; l")"; ) );
declare-binop( $">=", raw-type(U32), raw-type(U32), raw-type(U64), ( l"("; x; l">="; y; l")"; ) );

declare-binop( $"+", raw-type(U32), raw-type(U32), raw-type(U32), ( l"("; x; l"+"; y; l")"; ) );
declare-binop( $"-", raw-type(U32), raw-type(U32), raw-type(U32), ( l"("; x; l"-"; y; l")"; ) );
declare-binop( $"^", raw-type(U32), raw-type(U32), raw-type(U32), ( l"("; x; l"^"; y; l")"; ) );
declare-binop( $"/", raw-type(U32), raw-type(U32), raw-type(U32), ( l"("; x; l"/"; y; l")"; ) );
declare-binop( $"%", raw-type(U32), raw-type(U32), raw-type(U32), ( l"("; x; l"%"; y; l")"; ) );
declare-binop( $"*", raw-type(U32), raw-type(U32), raw-type(U32), ( l"("; x; l"*"; y; l")"; ) );

declare-unop( not, raw-type(U32), raw-type(U64), ( l"(!"; x; l")"; ) );

declare-unop( into-branch-conditional, raw-type(U32), raw-type(BranchConditional), x );

let cmp(l: U32, r: U32): Ord = if l==r then Equal() else if l<r then LessThan() else GreaterThan();

let print(io: IO::File, x: U32): Nil = fprintf(io, c"%u", x);
