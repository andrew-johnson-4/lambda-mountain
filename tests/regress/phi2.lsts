
import LIB/default.lsts;

type Phi2T = { id:U64 };

let del(t: Phi2T): Nil = (
   print("destroyed \{t.id}\n");
);

let f(x: Phi2T): Nil = ();  # end of lifetime
let g(x: Phi2T): Phi2T = x; # transfer of lifetime

if true {
   let binding = Phi2T(1);  # transfer of lifetime to scope
   Phi2T(2);                # end of lifetime
   f(Phi2T(3));             # transfer of lifetime to function
   g(Phi2T(4));             # transfer of lifetime to function
};
