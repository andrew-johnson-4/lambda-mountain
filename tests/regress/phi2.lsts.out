destroyed 2
destroyed 3
destroyed 4
destroyed 1
